ns for fx: bug fix for shadowing fp16 emulation patterns

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56927 ns for fx: add missing add_relu and mul_relu patterns
* #56762 ns for fx: additional bugfix for user defined functions
* #56742 ns for fx: allow comparing int8 to int8 for functionals
* #56493 ns for fx: add option to skip matching classes and functions
* #56408 ns for fx: support binary ops when adding unshadowed loggers for inputs
* **#56384 ns for fx: bug fix for shadowing fp16 emulation patterns**
* #56311 ns for fx: add fp16 function shadowing
* #56301 ns for fx: allow user functions in shadowing
* #56296 ns for fx: move node I/O dtype mapping to be local instead of global
* #56292 ns for fx: handling for user functions in weight and unshadowed act APIs
* #56283 ns for fx: enable user defined functions for graph matching

Summary:

Enables shadow copies of fp16 emulation patterns where weights
are cast to fp16 before being passed to linear.  This previously
did not work because copying of `call_method` nodes was not implemented.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16_vs_linear_fp16_shadow_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27857735](https://our.internmc.facebook.com/intern/diff/D27857735)

vkuzo committed:
ns for fx: bug fix for shadowing fp16 emulation patterns

Summary:

Enables shadow copies of fp16 emulation patterns where weights
are cast to fp16 before being passed to linear.  This previously
did not work because copying of `call_method` nodes was not implemented.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16_vs_linear_fp16_shadow_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -220,6 +220,46 @@ def _insert_dtype_cast_after_node(
     else:
         raise AssertionError(f"type f{type(prev_node_c)} is not handled")
 
+def _copy_node_from_a_to_c(
```
I don't have much context, a n00b question, why can't we use node_copy api?

vkuzo replied:
thanks, didn't know about that.  Wrote down a TODO to take a look in the near future.  Most likely we can change it, will just need to retest, hoping to do that in a future PR.

vkuzo committed:
Update on "ns for fx: bug fix for shadowing fp16 emulation patterns"


Summary:

Enables shadow copies of fp16 emulation patterns where weights
are cast to fp16 before being passed to linear.  This previously
did not work because copying of `call_method` nodes was not implemented.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16_vs_linear_fp16_shadow_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27857735](https://our.internmc.facebook.com/intern/diff/D27857735)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: bug fix for shadowing fp16 emulation patterns"


Summary:

Enables shadow copies of fp16 emulation patterns where weights
are cast to fp16 before being passed to linear.  This previously
did not work because copying of `call_method` nodes was not implemented.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16_vs_linear_fp16_shadow_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27857735](https://our.internmc.facebook.com/intern/diff/D27857735)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: bug fix for shadowing fp16 emulation patterns"


Summary:

Enables shadow copies of fp16 emulation patterns where weights
are cast to fp16 before being passed to linear.  This previously
did not work because copying of `call_method` nodes was not implemented.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16_vs_linear_fp16_shadow_activations
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27857735](https://our.internmc.facebook.com/intern/diff/D27857735)

[ghstack-poisoned]
